include/exadg/functions_and_boundary_conditions/interface_coupling.h

```diff
@@ -533,175 +142,70 @@ class InterfaceCoupling
+          return marked_vertices;
+        }));
+
+      map_communicator[quadrature].reinit(array_q_points_dst,
+                                          dof_handler_src_in.get_triangulation(),
+                                          mapping_src);
     }
 
     // finally, give boundary condition access to the data
     for(auto boundary : map_bc)
     {
       boundary.second->set_data_pointer(map_index_dst, map_solution_dst);
     }
+
+    update_data(dof_vector_src_in); // TODO: needed?
   }
 
   void
   update_data(VectorType const & dof_vector_src)
   {
-    VectorTypeDouble         dof_vector_src_double_copy;
-    VectorTypeDouble const * dof_vector_src_double_ptr;
-    if(std::is_same<double, Number>::value)
-    {
-      dof_vector_src_double_ptr = reinterpret_cast<VectorTypeDouble const *>(&dof_vector_src);
-    }
-    else
-    {
-      dof_vector_src_double_copy = dof_vector_src;
-      dof_vector_src_double_ptr  = &dof_vector_src_double_copy;
-    }
+    dof_vector_src.update_ghost_values();
 
-    dof_vector_src_double_ptr->update_ghost_values();
-
-    /*
-     * 1. Interpolate data on src-side
-     */
     for(auto quadrature : quad_rules_dst)
     {
-      ArrayVectorTensor & array_solution_dst = map_solution_dst.find(quadrature)->second;
-
-      std::map<mpi_rank, ArrayVectorCache> & mpi_map_cache_src =
-        map_cache_src.find(quadrature)->second;
-      std::map<mpi_rank, ArrayVectorTensor> & mpi_map_solution_src =
-        map_solution_src.find(quadrature)->second;
-
-      for(auto it : mpi_map_cache_src)
-      {
-        mpi_rank const      proc               = it.first;
-        ArrayVectorCache &  array_cache_src    = mpi_map_cache_src.find(proc)->second;
-        ArrayVectorTensor & array_solution_src = mpi_map_solution_src.find(proc)->second;
-
-        for(types::global_dof_index q = 0; q < array_cache_src.size(); ++q)
-        {
-          std::vector<Cache> &                     vector_cache    = array_cache_src[q];
-          std::vector<Tensor<rank, dim, double>> & vector_solution = array_solution_src[q];
-
-          // interpolate solution from dof vector using cached data
-          for(unsigned int i = 0; i < vector_cache.size(); ++i)
-          {
-            vector_solution[i] = Interpolator<rank, dim, double>::value(*dof_handler_src,
-                                                                        *dof_vector_src_double_ptr,
-                                                                        vector_cache[i].first,
-                                                                        vector_cache[i].second);
-          }
-        }
-      }
-
-      /*
-       * 2. Communication: transfer results back to dst-side
-       */
-      InterfaceCommunicator<dim, dim> & communicator = map_communicator.find(quadrature)->second;
-
-      communicator.process(mpi_map_solution_src, array_solution_dst, 11 + quadrature);
+      const auto result =
```

Comment:
Do you think it would make sense to spell out the type here (I guess it depends whether we have a good typedef otherwise it might get too verbose)? Also, we should do
```suggestion
      auto const result =
```

Reply:
const as a trailing keyword is defined in our ExaDG guidelines in the wiki. So I agree. I comment on the other aspect somewhere else.